Fix UTF32toUTF8 will produce invalid transition

tang-hi opened the pull request:
FIX ISSUE #12458

tang-hi committed:
fix error convert from utf32 to utf8

tang-hi committed:
fix utf8 error

tang-hi committed:
fix typo

mikemccand commented on lucene/core/src/java/org/apache/lucene/util/automaton/UTF32ToUTF8.java:
```diff
@@ -238,6 +238,10 @@ private void end(int start, int end, UTF8Sequence endUTF8, int upto, boolean doA
         // doesn't accept certain byte sequences) -- there
         // are other cases we could optimize too:
         startCode = 194;
```
Let's maybe fix this one to hex as well (0xC2)?

mikemccand commented on lucene/core/src/java/org/apache/lucene/util/automaton/UTF32ToUTF8.java:
```diff
@@ -238,6 +238,10 @@ private void end(int start, int end, UTF8Sequence endUTF8, int upto, boolean doA
         // doesn't accept certain byte sequences) -- there
         // are other cases we could optimize too:
```
Is this comment (`there are other cases we could optimize too`) still true :)  Or are these two new ifs covering them AND fixing this sneaky bug?

mikemccand commented on lucene/core/src/test/org/apache/lucene/util/TestUnicodeUtil.java:
```diff
@@ -188,6 +191,30 @@ public void testUTF8CodePointAt() {
     }
   }
 
+  public void testUTF8TwoToThreeBytes() throws Exception {
```
Maybe also add the three-to-four and one-to-two cases?

mikemccand reviewed (commented):
Thanks @tang-hi -- what a sneaky bug!  It's crazy it took randomized testing so long to eek out this failure, and also wonderful that it did.

How do we characterize the end user impact of this bug?  We need a well written CHANGES entry.  The impact should be small, since it's accepting illegal UTF-8 sequences?  It'd only affect users operating entirely in binary space (not Unicode encoded UTF-8 terms, which is the vast majority of users / default case).

tang-hi committed:
In the previous versions of Lucene, an incorrect assumption was made that a codepoint would start at 0x80 when the number of bits is 6.
This has now been corrected. The correct behavior is that when the length of the codepoint is 3 and the first byte is 0xE0, it will start from 0xA0.
Similarly, when the length of the codepoint is 4 and the first byte is 0xF0, it will start from 0x90.

tang-hi commented:
@mikemccand Thanks for the review. I have already made the updates to the pull request based on your review comments. 😄

tang-hi committed:
fix typo

mikemccand commented:
> @mikemccand Thanks for the review. I have already made the updates to the pull request based on your review comments. 😄

Wow that was fast, thank you!  I'll try to review soon -- I gotta stare at this long enough to understand the bug / fix / UTF-8 details.  Thank you for digging on this tricky case so quickly.

tang-hi commented:
2 byte -> 3 byte
<img width="749" alt="image" src="https://github.com/apache/lucene/assets/72755185/6202d3d8-7c87-42cf-9ba3-ef574260ba5d">

3 byte -> 4 byte
<img width="714" alt="image" src="https://github.com/apache/lucene/assets/72755185/6ba4725b-fb83-469d-a200-7801ffb21cd0">

hope these pictures will be helpful

mikemccand commented on lucene/core/src/java/org/apache/lucene/util/automaton/UTF32ToUTF8.java:
```diff
@@ -232,12 +232,18 @@ private void end(int start, int end, UTF8Sequence endUTF8, int upto, boolean doA
     } else {
+      // There are three special case
+      // 1. if codepoint's numBits is 5, byte start from 0xC2
+      // 2. if codepoint's len is 3  and  first byte is 0xE0. the second byte start from 0xA0
```
Extra space before `and` and `first`?

mikemccand commented on lucene/core/src/java/org/apache/lucene/util/automaton/UTF32ToUTF8.java:
```diff
@@ -232,12 +232,18 @@ private void end(int start, int end, UTF8Sequence endUTF8, int upto, boolean doA
-        startCode = 194;
+        startCode = 0xC2;
+      } else if (endUTF8.len == 3 && upto == 1 && endUTF8.byteAt(0) == 0xE0) {
+        startCode = 0xA0;
```
Can you move the above three bullets into each of these internal `if`s to make the special casing clearer?

E.g. in this `if` maybe say:

```
// the first length=3 UTF8 Unicode character is E0 A0 80 so we must special case A0 as the 2nd byte when E0 was the first byte of endUTF8 in this case
```

mikemccand reviewed (commented):
Phew, I finally understand this bug!

And I realize how I messed this up long ago after studying the UTF-8 encoding table ... from the line for the first 3-byte UTF-8 sequence:

```
U+0800 | U+FFFF | 1110xxxx | 10xxxxxx | 10xxxxxx
```

one cannot assume that the `xxxx` of the not-first UTF-8 byte will be 0s!  For the 2->3 and 3->4 case they are not :)

Thank you @tang-hi for bearing with me as I remember this tricky stuff.  I left a bunch of small comments to help readability maybe.  I think the fix is fundamentally correct!

mikemccand commented on lucene/core/src/test/org/apache/lucene/util/automaton/TestStringsToAutomaton.java:
```diff
@@ -142,22 +141,11 @@ private void checkAutomaton(List<BytesRef> expected, Automaton a, boolean isBina
     }
 
     // Make sure every term produced by the automaton is expected
-    FiniteStringsIterator it = new FiniteStringsIterator(a);
```
Hmm why did you remove the `isBinary` case for this test?

tang-hi replied:
Because it was added by @gsmiller in https://github.com/apache/lucene/pull/12461 for the purpose of passing the CI test, it should be restored since we have already fixed the bug.

mikemccand commented on lucene/core/src/test/org/apache/lucene/util/TestUnicodeUtil.java:
```diff
@@ -188,6 +191,37 @@ public void testUTF8CodePointAt() {
+    b.addTransition(s1, s2, 0x800);
+    b.addTransition(s1, s2, 0xFFFF);
+    // utf8 codepoint length is 4
+    b.addTransition(s1, s2, 0x10000);
```
I'm pretty sure the Automaton builder collapses adjacent transitions like this, but for paranoia, could you add the range explicitly?  E.g.:

```
b.addTransition(s1, s2, 0x7f, 0x80);
```

(Instead of two separate `addTransition` calls).  And same for the other two (four) transitions?

tang-hi committed:
update comment and improve readability based one review comments

tang-hi committed:
fix typo

tang-hi commented:
I have already updated the PR. You may review it at your convenience.

gsmiller commented:
Thanks @tang-hi for this change! I'll also try to have a look at this tomorrow or Friday since I already spent some time untangling the code in here :)

gsmiller commented on lucene/CHANGES.txt:
```diff
@@ -182,6 +182,10 @@ Bug Fixes
 * GITHUB#12451: Change TestStringsToAutomaton validation to avoid automaton conversion bug discovered in GH#12458
   (Greg Miller).
 
+* GITHUB#2472: UTF32ToUTF8 would sometimes accept extra invalid UTF-8 binary sequences.  This should not have any
```
Unless a user was directly using `UTF32ToUTF8#convert` right? Maybe you could mention that somehow?

gsmiller commented on lucene/core/src/java/org/apache/lucene/util/automaton/UTF32ToUTF8.java:
```diff
@@ -35,12 +35,12 @@ public final class UTF32ToUTF8 {
     int v = 2;
-    for (int i = 0; i < 32; i++) {
-      MASKS[i] = v - 1;
+    for (int i = 0; i < 7; i++) {
```
Nice tidy-up!

gsmiller commented on lucene/core/src/java/org/apache/lucene/util/automaton/UTF32ToUTF8.java:
```diff
@@ -227,19 +227,24 @@ private void end(int start, int end, UTF8Sequence endUTF8, int upto, boolean doA
-        // doesn't accept certain byte sequences) -- there
-        // are other cases we could optimize too:
-        startCode = 194;
+      if (endUTF8.len == 2) {
```
Maybe we should reference the issue number in a comment here that tracked the bug for future readers? It's a pretty tricky bug, so having the issue number in a comment might be useful?

gsmiller commented on lucene/core/src/java/org/apache/lucene/util/automaton/UTF32ToUTF8.java:
```diff
@@ -227,19 +227,24 @@ private void end(int start, int end, UTF8Sequence endUTF8, int upto, boolean doA
+        // the first length=3 UTF8 Unicode character is E0 A0 80,
+        // so we must special case 0xA0 as the 2nd byte when E0 was the first byte of endUTF8.
+        startCode = 0xA0;
+      } else if (endUTF8.len == 4 && upto == 1 && endUTF8.byteAt(0) == 0xF0) {
```
I'm sort of contorting my brain to try to figure out if this can ever actually happen since the max UTF8 length is 4 bytes. I'm not opposed to keeping it, but for my own sanity, did you actually trip this problem with a test case or see something I'm overlooking?

tang-hi replied:
If you comment it out, the test `testUTF8SpanMultipleBytes` in `TestUnicodeUtil.java` will fail. This is because when there is a transition span from 0xFFFF (3 bytes) to 0x10000 (4 bytes), it will produce an incorrect result.

gsmiller replied:
Right, of course. That makes sense. My brain was a bit tired at the end of the day yesterday when I was looking through this, and I had an "off by one" bug.

gsmiller reviewed (commented):
Pretty elegant fix for a tricky bug. Thank you!

tang-hi committed:
upadte code based on code reivew

tang-hi commented:
Hi, @gsmiller! I have already updated the PR. You may review it at your convenience.

gsmiller commented:
Thanks @tang-hi. This LGTM now (I have two outstanding, small bits of feedback/questions above if you have a chance, but they're not blocking). I'll leave this open for another day incase Mike has any additional feedback, but will otherwise get it merged in a day or so. Thanks again!

gsmiller reviewed (approved):
LGTM! Will merge tomorrow unless Mike has additional feedback.

gsmiller commented:
Merged and back-ported. Thanks again @tang-hi! Nice fix for a reasonably tricky issue!
